README


This archive contains the distribution Catalyst-Plugin-Log-Log4perlSimple,
version 0.2:

  Simple Log4perl plugin for Catalyst

This software is copyright (c) 2011 by Martyn Smith.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Catalyst-Plugin-Log-Log4perlSimple

0.2       2011-06-15 15:03:49 Pacific/Auckland

0.1       2011-06-15 13:28:46 Pacific/Auckland


lib/Catalyst/Plugin/Log/Log4perlSimple.pm
package Catalyst::Plugin::Log::Log4perlSimple;
BEGIN {
  $Catalyst::Plugin::Log::Log4perlSimple::VERSION = '0.2';
}

# ABSTRACT: Simple Log4perl plugin for Catalyst

use 5.010;
use strict;
use warnings;
use Moose;
use namespace::autoclean;
use Catalyst::Log::Log4perlSimple;

=head1 NAME

Catalyst::Plugin::Log::Log4perlSimple

=head1 SYNOPSIS

  # in MyApp.pm

  use Catalyst qw( Log::Log4perlSimple );

  # in myapp.conf

  # note that this configuration is entirely optional. The block below is
  # indicating the default values for everything, so if they look okay to you,
  # just omit the configuration entirely.

  <Plugin Log::Log4perlSimple>

      # Set this to 0 or 1 to indicate if you would like Catalyst debugging output.
      catalyst_debug 0

      # Set this to 0 or 1 to indicate if you would like Catalyst statistics output.
      catalyst_stats 0

      # What is the lowest level of debugging information you would like output
      # by by Log4perl (trace, debug, info, or warn)
      log_level debug

      # Boolean to control if we want to log to screen
      screen 1

      # Optional control specifying a filename to write log data to (comment this
      # out to disable writing to a file)
      #file /path/to/somefile.log

  </Plugin>

=head1 DESCRIPTION

Provides a zero configuration alternative to L<Catalyst::Log>.

Instantly gives you coloured terminal output and timestamps on your development
server.

Provides a trivial mechanism for routing log messages to a file (configurable
via your application's config file).

=head1 AUTHOR

Martyn Smith <martyn@dollyfish.net.nz>

=head1 METHODS

=head2 setup()

Implementation of the setup callback for L<Catalyst> plugins. This is used to
setup up the logging object.

=cut

sub setup {
    my ($class) = @_;

    my $defaults = {
        catalyst_debug => 0,
        catalyst_stats => 0,
        log_level => 'debug',
        screen => 1,
        file => undef,
    };

    my $config = $class->config->{'Plugin::Log::Log4perlSimple'};
    $config = {} unless UNIVERSAL::isa($config, 'HASH');

    $class->log(Catalyst::Log::Log4perlSimple->new);

    foreach my $key ( keys %{$defaults} ) {
        $config->{$key} = $defaults->{$key} unless exists $config->{$key};
    }

    if ( $config->{catalyst_debug} ) {
        Class::MOP::get_metaclass_by_name($class)->add_method('debug' => sub { 1 });
    }

    if ( $config->{catalyst_stats} ) {
        Class::MOP::get_metaclass_by_name($class)->add_method('use_stats' => sub { 1 });
    }

    if ( $config->{screen} ) {
        $class->log->screen_output(1);
    }

    if ( $config->{file} ) {
        $class->log->file_output($config->{file});
    }

    $class->maybe::next::method(@_);
};

1;


lib/Log/Log4perl/Appender/CatalystProxy.pm
package Log::Log4perl::Appender::CatalystProxy;
BEGIN {
  $Log::Log4perl::Appender::CatalystProxy::VERSION = '0.2';
}

our @ISA = qw(Log::Log4perl::Appender);

use 5.010;
use warnings;
use strict;

use Log::Log4perl::Level;

=head1 NAME

Log::Log4perl::Appender::CatalystProxy

=head1 SYNOPSIS

Generally speaking you probably don't want to use this directly, instead you
should use L<Catalyst::Plugin::Log::Log4perlSimple>.

If you do want do use it, refer to the L<Log::Log4perl> documentation.

=head1 DESCRIPTION

This is a custom appender for L<Log::Log4perl> that proxies messages to other
appenders. It is capable of pushing messages out to multiple appenders, and
also of buffering messages until an appender is allocated to it.

=head1 AUTHOR

Martyn Smith <martyn@catalyst.net.nz>

=head1 METHODS

=head2 new(%options)

Create an instance of this appender.

Options are:
    
- autoflush (boolean that indicates if logged messages should be proxied immediately)
- appenders (arrayref of appenders to proxy to; defaults to [])

=cut

sub new {
    my ($class, @options) = @_;

    my $self = {
        name      => 'unknown name',
        buffer    => [],
        autoflush => 1,
        appenders => [],
        @options,
    };

    return bless $self, $class;
};

=head2 log(%params)

Implementation of the log method for L<Log::Log4perl::Appender>

=cut

sub log {
    my ($self, %params) = @_;

    if ( $self->{autoflush} and @{$self->{appenders}} ) {
        foreach my $appender ( @{$self->{appenders}} ) {
            $appender->log(%params);
        }
    }
    else {
        push @{$self->{buffer}}, \%params;
    }
}

=head2 appenders()

Return a list of the appenders this object it proxying to

=cut

sub appenders { return @{shift->{appenders}} }

=head2 add_appender($appender)

Add $appender to the list of appenders this object is proxying for

=cut

sub add_appender {
    my ($self, $appender) = @_;

    push @{$self->{appenders}}, $appender;

    $self->flush if $self->{autoflush};
};

=head2 remove_appender($appender)

Remove $appender from the list of appenders this object is proxying for

=cut

sub remove_appender {
    my ($self, $appender) = @_;

    @{$self->{appenders}} = grep { $_ != $appender } @{$self->{appenders}};
}

=head2 flush()

Push everything in the buffer out to configured appenders then empty the
buffer.

=cut

sub flush {
    my ($self) = @_;

    return unless @{$self->{appenders}};

    foreach my $message ( @{$self->{buffer}} ) {
        foreach my $appender ( @{$self->{appenders}} ){
            $appender->log(%{$message});
        }
    }
    $self->{buffer} = [];
}

=head2 clear()

Clear the buffer without pushing the messages out to configured appenders

=cut

sub clear {
    my ($self) = @_;

    $self->{buffer} = [];
}

1;


Makefile.PL

use strict;
use warnings;

BEGIN { require 5.010; }

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Simple Log4perl plugin for Catalyst',
  'AUTHOR' => 'Martyn Smith <martyn@dollyfish.net.nz>',
  'BUILD_REQUIRES' => {
    'Test::More' => '0.88'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'Catalyst-Plugin-Log-Log4perlSimple',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Catalyst::Plugin::Log::Log4perlSimple',
  'PREREQ_PM' => {
    'Log::Log4perl' => '1.26',
    'Moose' => '0.94'
  },
  'VERSION' => '0.2',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



